File: homework/hw4/hw4_dev2.f90
!Template code for hw4
!Should be compiled with network.f90
module netstats
	use network
	use omp_lib

	contains

subroutine stats(n0,l,nt,m,qnetm,qmaxm,qvarm)
	!Input: 
	!n0,nl,nt: recursive network model parameters
	!m: number of network realizations to compute
	!Output:
	!qnetm: node lists for all m networks
	!qvarm: ensemble averages of var(q)
	!qmaxm: maximum qmax from m network realizations
	implicit none
	integer, intent(in) :: n0,l,nt,m
	integer, dimension(n0+nt,m), intent(out) :: qnetm
	integer, intent(out) :: qmaxm
	real(kind=8), intent(out) :: qvarm
	integer, dimension(n0+nt) :: qnet
	integer, dimension(n0+l*nt,2) :: enet
	real(kind=8) :: qmean
	real(kind=8), dimension(m) :: qvar
	integer :: i, qmax
	qmaxm = 0
        do i = 1,m
            call generate(n0,l,nt,qmax,qnet,enet)
            qnetm(:,i) =  qnet
            qmean = sum(qnet)/(n0+nt)
            qvar(i) = sum((qnet-qmean)**2)/(n0+nt)
            qmaxm = max(qmaxm,qmax)
        end do 
        qvarm = sum(qvar)/m
        
end subroutine stats


subroutine stats_omp(n0,l,nt,m,numThreads,qnetm,qmaxm,qvarm)
	!Input: 
	!n0,nl,nt: recursive network model parameters
	!m: number of network realizations to compute
	!numThreads: number of threads for parallel computation
	!Output:
	!qnetm: node lists for all m networks
	!qvarm: ensemble averages of var(q)
	!qmaxm: maximum qmax from m network realizations
	implicit none
	integer, intent(in) :: n0,l,nt,m,numThreads
	integer, dimension(n0+nt,m), intent(out) :: qnetm
	integer, intent(out) :: qmaxm
	real(kind=8), intent(out) :: qvarm
	real(kind=8), dimension(m) :: qvar
	real(kind=8) :: qmean
	integer, dimension(n0+nt) :: qnet
	integer, dimension(n0+l*nt,2) :: enet
	integer :: i, qmax
	!$call omp_set_num_threads(numThreads)
	qmaxm = 0
	!$OMP parallel do private(qmax,qmean)
        do i = 1,m
            call generate(n0,l,nt,qmax,qnetm(:,i),enet)
            qmean = sum(qnetm(:,i))/(n0+nt)
            qvar(i) = sum((qnetm(:,i)-qmean)**2)/(n0+nt)
            qmaxm = max(qmaxm,qmax)
        end do 
        !$OMP end parallel do 
        qvarm = sum(qvar)/m


end subroutine stats_omp


subroutine test_stats_omp(n0,l,nt,m,numThreads,walltime)
	!Input: same as stats_omp
	!Output: walltime: time for 100 calls to stats_par
	implicit none
	integer, intent(in) :: n0,l,nt,m,numThreads
	real(kind=8), intent(out) :: walltime
	integer, dimension(n0+nt,m) :: qnetm
	integer :: qmaxm,i
	real(kind=8), dimension(100) :: wtime
	real(kind=8) :: qvarm
	integer :: start,finish,clockrate
        if (numThreads>1) then 
            print *, 'parallel'
            do i=1,100
                call system_clock(start)
                call stats_omp(n0,l,nt,m,numThreads,qnetm,qmaxm,qvarm)
                call system_clock(finish,clockrate)
                wtime(i) = float(finish-start)/float(clockrate)
            end do
            walltime = sum(wtime)/100
        else 
            print *, 'normal'
            do i=1,100 
                call system_clock(start)
                call stats(n0,l,nt,m,qnetm,qmaxm,qvarm)
                call system_clock(finish,clockrate)
                wtime(i) = float(finish-start)/float(clockrate)
            end do
            walltime = sum(wtime)/100
        end if 

end subroutine test_stats_omp


end module netstats

File: homework/hw4/hw4_dev2.py
"""template code for hw4
"""
import numpy as np
import matplotlib.pyplot as plt
from ns import netstats as ns #assumes netstats module has been compiled as ns.so with f2py
import time

def convergence(n0,l,nt,m,numthreads,display=False):
    """test convergence of qmax_ave"""
    qmax_ave = np.zeros(m)
    M = range(1,m+1)
    for i in M:
        qnetm,qmaxm,qvarm = ns.stats(n0,l,nt,i)
        qmax = qnetm.max(axis=0)
        qmax_ave[i-1]= sum(qmax)/float(len(qmax))
    a,b = np.polyfit(np.log(M),np.log(qmax_ave),1)
    fit = a*np.log(M) + b
    k = -a
    if (display==True):
        plt.figure()
        plt.loglog(M,qmax_ave)
	plt.loglog(M,np.exp(fit))
    return k
	     
	
def speedup(n0,l,ntarray,marray):
    """test speedup of stats_par"""
    speedup = np.zeros((np.size(ntarray),np.size(marray)))
    for i in range(np.size(ntarray)):
        for j in range(np.size(marray)):
            nt = ntarray[i]
            m = marray[j]
            t1 = ns.test_stats_omp(n0,l,nt,m,1)
            t2 = ns.test_stats_omp(n0,l,nt,m,2)
            speedup[i,j] = t1/t2
    plt.figure()
    plt.plot(ntarray,speedup[:,0])
    plt.figure()
    plt.plot(marray,speedup[0,:])
        
	


def degree_average(n0,l,nt,m,display=False):
	"""compute ensemble-averaged degree distribution"""	
        



#if __name__ == '__main__':
    #ntarray = np.array([5,10,50,100,500,1000])
    #marray = np.array([5,10,50,100,500,1000])